src/tink/macro/TypeMap.hx

```diff
@@ -25,9 +25,17 @@ class TypeMap<V> extends BalancedTree<Type, V> implements IMap<Type, V> {
     
     return switch k1.getIndex() - k2.getIndex() {
       case 0: 
-        Reflect.compare(k1.toString(), k2.toString());//much to my surprise, this actually seems to work (at least with 3.4)
+        Reflect.compare(stringify(k1), stringify(k2));//much to my surprise, this actually seems to work (at least with 3.4)
       case v: v;
     }
   }
   
+  static var NESTED_NULL_REGEX = ~/Null<Null<([^>]*)>>/g; // a hack to flatten nested Null<T>
+  function stringify(type:Type) {
+    var s = type.toString();
+    while(NESTED_NULL_REGEX.match(s))
+      s = NESTED_NULL_REGEX.replace(s, "Null<$1>");
```

Comment:
This one gives me a bit of a headache. I really hate fumbling with strings in any manner, but regexes only make it worse, because regular grammars do not cover context free languages. Cases like `Null<Null<{ o: Array<Array<Int>> }>>` do not work: https://try.haxe.org/#FD6Fd

Because the goal is not to construct a syntactically valid string, but a key, one might get away with something like this. But as shown in the example, the current approach produces different keys for types that should be considered the same.